File: java/SantaClaus.java
import java.util.Random;
import java.util.concurrent.Semaphore;

//Solution taken from little book of semaphores and applied to java
public class SantaClaus {

    static Semaphore mutex = new Semaphore(1);
    static Semaphore elfTex = new Semaphore(1);
    static Semaphore santaSem = new Semaphore(0);
    static Semaphore reindeerSem = new Semaphore(0);
    static int maxReindeer = 9;
    static int maxElves = 100;
    static int elves = 0;
    static int reindeer = 0;

    public static void sleepThread(int time) {
        try {
            Thread.sleep((long) (Math.random() * time));
        } catch (InterruptedException e) {
            System.out.println(e.getMessage());
        }
    }

    public static void main(String args[]) {
        Elf elf = new Elf();
        Reindeer reindeer = new Reindeer();
        Santa santa = new Santa();

        // init santa
        Thread santaThread = new Thread(santa);
        santaThread.setName("santa");
        santaThread.start();

        // init reindeer
        Thread[] reindeers = new Thread[maxReindeer];
        for (int i = 0; i < reindeers.length; i++) {
            reindeers[i] = new Thread(reindeer);
            reindeers[i].setName("reindeer " + i);
            reindeers[i].setDaemon(true);
            reindeers[i].start();
        }

        // init elves
        Thread[] elves = new Thread[maxElves];
        for (int i = 0; i < elves.length; i++) {
            elves[i] = new Thread(elf);
            elves[i].setName("elf " + i);
            elves[i].setDaemon(true);
            elves[i].start();
        }
    }

    static class Elf implements Runnable {
        @Override
        public void run() {
            try {
                sleepThread(250);
                // Make sure no other elves are incrementing elves variable
                elfTex.acquire();
                mutex.acquire();
                elves++;
                // If max elves, notify santa otherwise move on
                if (elves == 3) {
                    santaSem.release();
                } else {
                    elfTex.release();
                }
                mutex.release();

                // Get santas help
                sleepThread(250);

                mutex.acquire();
                elves--;
                if (elves == 0) {
                    elfTex.release();
                }
                mutex.release();

            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    static class Santa implements Runnable {
        @Override
        public void run() {
            try {
                while (true) {
                    santaSem.acquire();
                    mutex.acquire();
                    // Wait for reindeer to arrive
                    if (reindeer >= maxReindeer) {
                        System.out.println("Reindeers finished, hitching sled!");
                        reindeerSem.release(9);
                        reindeer -= 9;
                        return;
                    } else if (elves == 3) {
                        System.out.println("Helping three elves");
                        sleepThread(250);
                    }
                    mutex.release();
                }

            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    static class Reindeer implements Runnable {
        @Override
        public void run() {
            try {
                sleepThread(2500);
                mutex.acquire();
                reindeer++;
                System.out.println("New reindeer arrived");
                if (reindeer == maxReindeer) {
                    System.out.println("Last reindeer arrived, notifying santa");
                    santaSem.release();
                }
                mutex.release();
                reindeerSem.acquire();
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }

        }
    }
}

File: golang/santaclaus.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"

	"github.com/pkg/profile"
)

var reindeer = 9
var elves = 100
var wg sync.WaitGroup

func sleepThread(counter int) {
	time.Sleep(time.Duration(rand.Intn(counter)) * time.Millisecond)
}

func main() {
	defer profile.Start().Stop()
	//Buffered to number of reindeers remaining
	reindeersRemaining := make(chan int, reindeer)
	initReindeer(reindeersRemaining)

	reindeersFinished := make(chan int)
	santaAvailable := make(chan int)
	threeElves := make(chan int)
	elvesWaiting := make(chan int, 3)

	//Wait for reindeer to finish
	go waitForReindeer(reindeersFinished)
	//Start santa helping elves
	go helpElves(santaAvailable, threeElves)

	wg.Add(1)
	//Start the elves at random times
	for i := 0; i < elves; i++ {
		go getHelp(santaAvailable, elvesWaiting, threeElves, i)
	}

	for i := 0; i < reindeer; i++ {
		go arriveReindeer(reindeersRemaining, reindeersFinished)
	}
	//Make sure we wait till all elves have finished or the last reindeer arrives
	wg.Wait()
}

func initReindeer(reindeersRemaining chan int) {
	//Initialize to 9 reindeers remaining
	for i := 0; i < 9; i++ {
		reindeersRemaining <- 1
	}
}

func arriveReindeer(reindeersRemaining, reindeersFinished chan int) {
	sleepThread(2500)
	<-reindeersRemaining
	fmt.Println("New reindeer arrived")
	if len(reindeersRemaining) == 0 {
		fmt.Println("Last reindeer arrived, notifying santa")
		reindeersFinished <- 1
	}
}

func waitForReindeer(reindeersFinished chan int) {
	//Unbufferred channel, blocks till reindeers finished
	<-reindeersFinished
	fmt.Println("Reindeers finished, hitching sled!")
	wg.Done()
	return
}

func helpElves(santaAvailable, threeElves chan int) {
	for {
		//Wait till three elves have arrived
		<-threeElves
		fmt.Println("Helping three elves")
		//Helping
		sleepThread(250)
		//Signal you are available again
		santaAvailable <- 1

	}
}

func getHelp(santaAvailable, elvesWaiting, threeElves chan int, id int) {
	sleepThread(250)
	//Go into the queue of elves waiting
	elvesWaiting <- 1
	//If you are third elf, notify santa
	if len(elvesWaiting) == 3 {
		for i := 0; i < 3; i++ {
			//Clear all elves waiting before signaling that three elves are groupe
			<-elvesWaiting
		}
		threeElves <- 1
	}
	<-santaAvailable
	//Get santas help
	sleepThread(250)
	//No longer waiting
}
